src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,241 @@
+use std::borrow::Cow;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check for objects returned from function calls in a match scrutinee that have the attribute
+    /// `has_significant_drop`.
+    ///
+    /// ### Why is this bad?
+    /// The attribute `has_significant_drop` is added to types whose Drop impls have an important
+    /// side-effect, such as unlocking a mutex, making it important for users to be able to
+    /// accurately understand their lifetimes. When a temporary is returned in a function call in a
+    /// match scrutinee, its lifetime lasts until the end of the match block, which may be
+    /// surprising. It can also lead to deadlocks if the match scrutinee uses a function call that
+    /// returns a MutexGuard and then tries to lock again in one of the match arms, because the
+    /// MutexGuard will not be dropped until the end of the match block and thus will not unlock.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     match mutex.lock().unwrap().foo() {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar(); // Deadlock!
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// use std::sync::Mutex;
+    ///
+    /// struct State {}
+    ///
+    /// impl State {
+    ///     fn foo(&self) -> bool {
+    ///         true
+    ///     }
+    ///
+    ///     fn bar(&self) {}
+    /// }
+    ///
+    /// fn main() {
+    ///     let mutex = Mutex::new(State {});
+    ///
+    ///     let is_foo = mutex.lock().unwrap().foo();
+    ///     match is_foo {
+    ///         true => {
+    ///             mutex.lock().unwrap().bar();
+    ///         }
+    ///         false => {}
+    ///     };
+    ///
+    ///     println!("All done!");
+    /// }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub SIGNIFICANT_DROP_IN_SCRUTINEE,
+    nursery,
+    "warns when a temporary of a type with a drop with a significant side-effect might have a surprising lifetime"
+}
+
+declare_lint_pass!(SignificantDropInScrutinee => [SIGNIFICANT_DROP_IN_SCRUTINEE]);
+
+impl<'tcx> LateLintPass<'tcx> for SignificantDropInScrutinee {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let Some((sig_temp_span, expr_span)) = has_significant_drop_in_scrutinee(cx, expr) {
+            let replacement = reindent_multiline(
+                Cow::Owned(format!("let value = {};\nmatch ", snippet(cx, expr_span, ".."))),
+                true,
+                Some(4),
```

Comment:
This hard-coded `4` only works, because all `match` expressions in the test are indented by exactly `4`. You should be able to use `indent_of` on the match expression to get the correct value to pass to this function.

https://github.com/rust-lang/rust/blob/ad46af24713115e7b9b258346e66b9b2d14eacfc/src/tools/clippy/clippy_utils/src/source.rs#L57-L65

Reply:
Thanks!